calendar/src/main/java/com/toggl/calendar/ui/CalendarLayoutCalculator.kt

```diff
@@ -0,0 +1,175 @@
+package com.toggl.calendar.ui
+
+import com.toggl.environment.services.time.TimeService
+import org.threeten.bp.Duration
+import org.threeten.bp.OffsetDateTime
+import org.threeten.bp.temporal.ChronoUnit
+import javax.inject.Inject
+import kotlin.math.absoluteValue
+
+typealias CalendarItemGroup = MutableList<Pair<Int, CalendarItem>>
+typealias CalendarItemGroups = MutableList<CalendarItemGroup>
+
+class CalendarLayoutCalculator @Inject constructor(private val timeService: TimeService) {
+
+    private val offsetFromNow: Duration = Duration.ofMinutes(7)
+    private val minimumDurationForUIPurposes: Duration = Duration.ofMinutes(15)
+
+    fun calculateLayoutAttributes(calendarItems: List<CalendarItem>): List<CalendarItemLayoutAttributes> {
+        if (calendarItems.isEmpty())
+            return emptyList()
+
+        return calendarItems
+            .mapIndexed { index, calendarItem -> index to calendarItem }
+            .sortedBy { (_, item) -> item.startTime() }
+            .fold(mutableListOf(), ::groupOverlappingItems)
+            .flatMap(::calculateLayoutAttributesFor)
+            .sortedBy { (index, _) -> index }
+            .map { (_, item) -> item }
+            .toList()
+    }
+
+    /**
+     * Aggregates the indexed calendar items into buckets. Each bucket contains the sequence of overlapping items.
+     * The items in a bucket don't overlap all with each other, but cannot overlap with items in other buckets.
+     *
+     * @param buckets   The list of agregated items
+     * @param indexedCalendarItem   The item to put in a bucket
+     * @return A list of buckets
+     */
+    private fun groupOverlappingItems(
+        buckets: CalendarItemGroups,
+        indexedCalendarItem: Pair<Int, CalendarItem>
+    ): CalendarItemGroups {
+        if (buckets.isEmpty()) {
+            buckets.add(mutableListOf(indexedCalendarItem))
+            return buckets
+        }
+
+        val now = timeService.now()
+        val group = buckets.last()
+        val maxEndTime = group.map { (_, item) -> calculateEndTimeWith(item, now) }.max() ?: OffsetDateTime.MIN
+        if (indexedCalendarItem.second.startTime() < maxEndTime)
+            group.add(indexedCalendarItem)
+        else
+            buckets.add(mutableListOf(indexedCalendarItem))
+
+        return buckets
+    }
+
+    /**
+     * Calculates the layout attributes for the indexed calendar items in a bucket.
+     * The calculation is done minimizing the number of columns.
+     *
+     * @param bucket
+     * @return An list of indexed calendar attributes
+     */
+    private fun calculateLayoutAttributesFor(bucket: CalendarItemGroup): List<Pair<Int, CalendarItemLayoutAttributes>> {
+        val left = bucket.filter { indexedItem -> indexedItem.second is CalendarItem.CalendarEvent }.toMutableList()
+        val right = bucket.filter { indexedItem -> indexedItem.second is CalendarItem.TimeEntry }.toMutableList()
```

Comment:
```suggestion
        val left = bucket.filter { (_, item) -> item is CalendarItem.CalendarEvent }.toMutableList()
        val right = bucket.filter { (_, item)  -> item is CalendarItem.TimeEntry }.toMutableList()
```

This is always preferable to `.first` or `.second` for readability reasons imo.

Reply:
Good suggestion. I'll apply everywhere else.